src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/WorkspaceProjectStateChangeDetector.cs

```diff
@@ -153,6 +163,49 @@ internal virtual void Workspace_WorkspaceChanged(object sender, WorkspaceChangeE
             }
         }
 
+        // Internal for testing
+        internal bool IsPartialComponentClass(Document document)
+        {
+            if (!document.TryGetSyntaxRoot(out var root))
+            {
+                return false;
+            }
+
+            var classDeclarations = root.DescendantNodes().OfType<ClassDeclarationSyntax>();
+            if (!classDeclarations.Any())
+            {
+                return false;
+            }
+
+            if (!document.TryGetSemanticModel(out var semanticModel))
+            {
+                // This will occasionally return false resulting in us not refreshing TagHelpers for component partial classes. This means there are situations when a users'
+                // TagHelper definitions will not immediately update but we will eventually acheive omniscience.
+                return false;
+            }
+
+            foreach (var classDeclaration in classDeclarations)
+            {
+                var classSymbol = semanticModel.GetDeclaredSymbol(classDeclaration) as INamedTypeSymbol;
+                if (classSymbol == null)
+                {
+                    continue;
+                }
+
+                // We can't access the compilation at this point so we need to do a "fuzzy" match for our IComponent base type. In the worst case scenario
+                // we'll end up triggering extra TagHelper updates for situations where a class implements some other IComponent type that looks like ours.
+                var implementsIComponent = classSymbol.AllInterfaces.FirstOrDefault(@interface => string.Equals(@interface.ToDisplayString(), ComponentsApi.IComponent.MetadataName)) != null;
```

Comment:
```suggestion
                var implementsIComponent = classSymbol.AllInterfaces.Any(@interface => string.Equals(@interface.ToDisplayString(), ComponentsApi.IComponent.MetadataName));
```

Reply:
Would it help to do a cursory check on the type name before doing `ToDisplayName()` e.g. `@interface.Name == nameof(ComponentsApi.IComponent) && @interface.ToDisplayString() == ComponentsApi.IComponent.MetadataName`

I don't think they cache the display strings, so this could quickly get expensive.